cmds/webboot/webboot.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/u-root/webboot/pkg/menu"
+)
+
+const (
+	downloadByLinkLabel = "Download by link"
+	tcURL               = "http://tinycorelinux.net/10.x/x86_64/release/TinyCorePure64-10.1.iso"
+)
+
+// bookmark record the list of iso webboot allow user to download
+var bookmark = map[string]DownloadedDistro{
+	"tinycore": DownloadedDistro{
+		isDefault: false,
+		url:       tcURL,
+		label:     "Download Tinycore v10.1",
+		isoName:   "TinyCorePure64-10.1.iso",
+		group:     "Tinycore",
+	},
+}
+
+// BootMedia contains required information of a bootable iso
+// no matter it is cached or downloaded
+type BootMedia interface {
+	Path() string
+	Name() string
+	// todo: may have Kernel(), Config() etc later
+}
+
+// DownloadedDistro contains information of the iso user want to download
+type DownloadedDistro struct {
```

Comment:
So looking at your structs, I still see a lot of similar fields. The goal with an interface is to reduce the duplication, right? Is there a way you can reduce the duplication here - for example, do you need the isDefault and label fields if you can already have methods for them from an interface?

Another question to consider is why DownloadedDistro needs its own struct. It seems like after the Exec() function executes, you won't use the url again, in which case there's no difference between that and CachedDistro, right? If I remember, you added this because you had 2 options, Download by bookmark and Download by link. Would it make sense to do something like this

```
type DownloadByBookmark struct{}

func (d *DownloadByBookmark) Label() {
    return "Download by bookmark"
}

func (d *DownloadByBookmark) Exec() {
    // Present a menu of the bookmarks, boot the one the user selects
}

type DownloadByLink struct{}

func (d *DownloadByLink) Label() {
    return "Download by link"
}

func (d *DownloadByLink) Exec() {
    // Ask for a link, download and boot
}
```
since you don't really need any information in the struct, since the user will be providing it to you?

Reply:
Get you, But then I still need another struct to implement the bookmark option.

Reply:
and I can not get a BootMedia interface for the download iso in the main program in this way right? because I have no attribute to store these information. All following process should be done in the Exec()

Reply:
Or we let the all following process happens Exec() anyway, and the BootMedia just a interface that provide Gabriel information for booting. So use this interface Gabriel could ignore the difference between cachediso and downloadediso

Reply:
I send a schematic diagram in the Chat